File: ReactAndroid/src/main/java/com/facebook/react/packagerconnection/JSPackagerClient.java
/**
 * Copyright (c) 2015-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the LICENSE file in the root
 * directory of this source tree. An additional grant of patent rights can be found in the PATENTS
 * file in the same directory.
 */

package com.facebook.react.packagerconnection;

import javax.annotation.Nullable;

import java.util.Map;

import com.facebook.common.logging.FLog;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import okhttp3.ws.WebSocket;

import org.json.JSONObject;

/**
 * A client for packager that uses WebSocket connection.
 */
final public class JSPackagerClient implements ReconnectingWebSocket.MessageCallback {
  private static final String TAG = JSPackagerClient.class.getSimpleName();
  private static final int PROTOCOL_VERSION = 2;

  public class Responder {
    private Object mId;

    public Responder(Object id) {
      mId = id;
    }

    public void respond(Object result) {
      try {
        JSONObject message = new JSONObject();
        message.put("version", PROTOCOL_VERSION);
        message.put("id", mId);
        message.put("result", result);
        mWebSocket.sendMessage(RequestBody.create(WebSocket.TEXT, message.toString()));
      } catch (Exception e) {
        FLog.e(TAG, "Responding failed", e);
      }
    }

    public void error(Object error) {
      try {
        JSONObject message = new JSONObject();
        message.put("version", PROTOCOL_VERSION);
        message.put("id", mId);
        message.put("error", error);
        mWebSocket.sendMessage(RequestBody.create(WebSocket.TEXT, message.toString()));
      } catch (Exception e) {
        FLog.e(TAG, "Responding with error failed", e);
      }
    }
  }

  public interface RequestHandler {
    public void onRequest(@Nullable Object params, Responder responder);
    public void onNotification(@Nullable Object params);
  }

  public static abstract class NotificationOnlyHandler implements RequestHandler {
    final public void onRequest(@Nullable Object params, Responder responder) {
      responder.error("Request is not supported");
      FLog.e(TAG, "Request is not supported");
    }
    abstract public void onNotification(@Nullable Object params);
  }

  public static abstract class RequestOnlyHandler implements RequestHandler {
    abstract public void onRequest(@Nullable Object params, Responder responder);
    final public void onNotification(@Nullable Object params) {
      FLog.e(TAG, "Notification is not supported");
    }
  }

  private ReconnectingWebSocket mWebSocket;
  private Map<String, RequestHandler> mRequestHandlers;

  public JSPackagerClient(String url, Map<String, RequestHandler> requestHandlers) {
    super();
    mWebSocket = new ReconnectingWebSocket(url, this);
    mRequestHandlers = requestHandlers;
  }

  public void init() {
    mWebSocket.connect();
  }

  public void close() {
    mWebSocket.closeQuietly();
  }

  @Override
  public void onMessage(ResponseBody response) {
    if (response.contentType() != WebSocket.TEXT) {
      FLog.w(
        TAG,
        "Websocket received message with payload of unexpected type " + response.contentType());
      return;
    }

    try {
      JSONObject message = new JSONObject(response.string());

      int version = message.optInt("version");
      String method = message.optString("method");
      Object id = message.opt("id");
      Object params = message.opt("params");

      if (version != PROTOCOL_VERSION) {
        FLog.e(
          TAG,
          "Message with incompatible or missing version of protocol received: " + version);
        return;
      }

      if (method == null) {
        abortOnMessage(id, "No method provided");
        return;
      }

      RequestHandler handler = mRequestHandlers.get(method);
      if (handler == null) {
        abortOnMessage(id, "No request handler for method: " + method);
        return;
      }

      if (id == null) {
        handler.onNotification(params);
      } else {
        handler.onRequest(params, new Responder(id));
      }
    } catch (Exception e) {
      FLog.e(TAG, "Handling the message failed", e);
    } finally {
      response.close();
    }
  }

  private void abortOnMessage(Object id, String reason) {
    if (id != null) {
      (new Responder(id)).error(reason);
    }

    FLog.e(TAG, "Handling the message failed with reason: " + reason);
  }
}

File: local-cli/server/util/jsPackagerClient.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

const WebSocket = require('ws');

const parseMessage = require('./messageSocket').parseMessage;

const PROTOCOL_VERSION = 2;
const TARGET_SERVER = 'server';

function getMessageId() {
  return `${Date.now()}:${Math.random()}`;
}

class JsPackagerClient {
  constructor(url) {
    this.ws = new WebSocket(url);
    this.msgCallbacks = new Map();

    this.openPromise = new Promise((resolve, reject) => {
      this.ws.on('error', error => reject(error));
      this.ws.on('open', resolve);
    });

    this.ws.on('message', (data, flags) => {
      const message = parseMessage(data, flags.binary);
      const msgCallback = this.msgCallbacks.get(message.id);
      if (message === undefined || message.id === undefined) {
        // gracefully ignore wrong messages or broadcasts
      } else if (msgCallback === undefined) {
        console.warn(`Response with non-existing message id: '${message.id}'`);
      } else {
        if (message.error === undefined) {
          msgCallback.resolve(message.result);
        } else {
          msgCallback.reject(message.error);
        }
      }
    });
  }

  sendRequest(method, target, params) {
    return this.openPromise.then(() => new Promise((resolve, reject) => {
      const messageId = getMessageId();
      this.msgCallbacks.set(messageId, {resolve: resolve, reject: reject});
      this.ws.send(
        JSON.stringify({
          version: PROTOCOL_VERSION,
          target: target,
          method: method,
          id: messageId,
          params: params,
        }),
        error => {
          if (error !== undefined) {
            this.msgCallbacks.delete(messageId);
            reject(error);
          }
        });
    }));
  }

  sendNotification(method, target, params) {
    return this.openPromise.then(() => new Promise((resolve, reject) => {
      this.ws.send(
        JSON.stringify({
          version: PROTOCOL_VERSION,
          target: target,
          method: method,
          params: params,
        }),
        error => {
          if (error !== undefined) {
            reject(error);
          } else {
            resolve();
          }
        });
    }));
  }

  sendBroadcast(method, params) {
    return this.sendNotification(method, undefined, params);
  }

  getPeers() {
    return new Promise((resolve, reject) => {
      this.sendRequest('getpeers', TARGET_SERVER, undefined).then(
        response => {
          if (!response instanceof Map) {
            reject('Results received from server are of wrong format:\n' +
                   JSON.stringify(response));
          } else {
            resolve(response);
          }
        },
        reject);
    });
  }

  getId() {
    return this.sendRequest('getid', TARGET_SERVER, undefined);
  }
}

module.exports = JsPackagerClient;
